core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -149,6 +155,67 @@ class FileMessageSet private[kafka](@volatile var file: File,
   }
 
   /**
+   * Search forward for the message whose timestamp is greater than or equals to the target timestamp.
+   *
+   * The search will stop immediately when it sees a message in format version before 0.10.0. This is to avoid
+   * scanning the entire log when all the messages are still in old format.
+   *
+   * @param targetTimestamp The timestamp to search for.
+   * @param startingPosition The starting position to search.
+   * @return None, if no message exists at or after the starting position.
+   *         Some(TimestampOffset) with (largest_timestamp_checked, the_next_offset_to_read) otherwise.
+   */
+  def searchForTimestamp(targetTimestamp: Long, startingPosition: Int): Option[TimestampOffset] = {
+    var maxTimestampChecked = Message.NoTimestamp
+    var lastOffsetChecked = -1L
+    val messagesToSearch = read(startingPosition, sizeInBytes)
+    for (messageAndOffset <- messagesToSearch) {
+      val message = messageAndOffset.message
+      lastOffsetChecked = messageAndOffset.offset
+      // Stop searching once we see message format before 0.10.0.
+      // This equivalent as treating message without timestamp has the largest timestamp.
+      // We do this to avoid scanning the entire log if no message has a timestamp.
+      if (message.magic == Message.MagicValue_V0)
+        return Some(TimestampOffset(maxTimestampChecked, messageAndOffset.offset))
```

Comment:
Yes, stopping early means we may not find the exact offset with a timestamp that we want. But it will not miss that message. 

In reality, a log segment may have messages w/ and w/o timestamp because some of the producers have been upgraded and others have not. What concerns me of not stopping when see old message is that if we don't have any message w/ a timestamp (or the timestamp we wanted locates close to the end of the log segment), we may end up with scanning the entire log segment. That will pollute the memory and impact the performance.

Since we are not appending a time index entry to the time index for the first message with a timestamp in a segment. Can we do the following?

1. In ```TimeIndex.lookup()```, when slotFor() returns -1 in the timeindex, we can just return the offset of the first entry instead of baseOffset (the offset will be baseOffset if the first message has a timestamp). This will essentially skip all the earlier messages w/o a timestamp in the log segment. If the time index is empty, we still return baseOffset. 

2. In ```searchForTimestamp()``` we keep searching when we see message w/o a timestamp until we find the timestamp. The only risk is that if two indexed offset in the time index are far away, we may still be scanning the entire log. For example, if message 0 has a timestamp 100, message 100000 has timestamp 200. When people are searching for timestamp 150, we will scan the message from 0 to 100000. But this should be rare.